python/unblob/handlers/executable/elf.py

```diff
@@ -29,8 +36,74 @@
 KERNEL_INIT_DATA_SECTION = ".init.data"
 
 
+# [Ref] https://github.com/upx/upx/blob/devel/src/stub/src/include/linux.h
+UPX_C_DEFINITIONS = r"""
+typedef struct packhead{
+    char magic[4];
+    uint8_t version;
+    uint8_t format;
+    uint8_t level;
+    uint8_t method;
+    uint64_t unknown1;
+    uint32_t u_filesize1;
+    uint32_t c_filesize;
+    uint32_t u_filesize2;
+    uint32_t unknown2;
+    uint32_t l_info_offset;
+} packhead_t;
+
+typedef struct l_info{
+    uint32_t l_checksum;
+    char l_magic[4];
+    uint16_t l_lsize;
+    uint8_t l_version;
+    uint8_t l_format;
+} l_info_t;
+"""
+upx_parser = StructParser(UPX_C_DEFINITIONS)
+
+
+def parse_upx_packhead(file: File):
+    return upx_parser.parse("packhead_t", file, Endian.LITTLE)
+
+
+def parse_upx_l_info(file: File):
+    return upx_parser.parse("l_info_t", file, Endian.LITTLE)
+
+
 @attrs.define(repr=False)
 class ElfChunk(ValidChunk):
+    def upx_checksum_validates(self, file: File, l_info, elf) -> bool:
+        size_pack2 = elf.last_offset_segment - l_info.l_lsize
+        size_aligment = round_up(size_pack2, 4)  # Forces to be mod 4
+        xct_off = any(section.name == "init" for section in elf.sections)
+        size_aligment += (4 & size_aligment) ^ (int(bool(xct_off)) << 2)  # 4 or 0
+        size_aligment += 8  # Added 2 times 4 byte (size of disp)
+        if xct_off:
+            size_aligment += 12
+        alignment = size_aligment - size_pack2
+        checksum_offset = elf.last_offset_segment - (l_info.l_lsize - alignment)
+        file.seek(checksum_offset, io.SEEK_SET)
+        adler32_checksum = 1
+        for chunk in iterate_file(file, checksum_offset, l_info.l_lsize - alignment):
+            adler32_checksum = zlib.adler32(chunk, adler32_checksum)
+        return adler32_checksum == l_info.l_checksum
+
+    def is_valid_upx(self, inpath: Path, elf) -> bool:
+        file = File.from_path(inpath)
+        try:
+            file.seek(-4, io.SEEK_END)  # last 4 bytes indicates where linfo ends
+            l_info_start_offset = convert_int32(file.read(4), Endian.LITTLE) - 12
+            file.seek(l_info_start_offset, io.SEEK_SET)
+            upx_header = parse_upx_l_info(file)
+            if upx_header.l_magic != b"UPX!":  # Magic
+                return False
+            if not self.upx_checksum_validates(file, upx_header, elf):
+                raise InvalidInputFormat("Invalid UPX checksum")
+            return True  # noqa: TRY300
+        except SeekError:
+            return False
```

Comment:
You can check the offsets rather than catching the `SeekError`:

```suggestion
        file.seek(-4, io.SEEK_END)  # last 4 bytes indicates where linfo ends
        l_info_start_offset = convert_int32(file.read(4), Endian.LITTLE) - 12
        if l_info_start_offset > file.size():
            return False
        file.seek(l_info_start_offset, io.SEEK_SET)
        upx_header = parse_upx_l_info(file)
        if upx_header.l_magic != b"UPX!":  # Magic
            return False
        if not self.upx_checksum_validates(file, upx_header, elf):
            raise InvalidInputFormat("Invalid UPX checksum")
        return True
```